src/commands/tail/mod.rs

```diff
@@ -0,0 +1,220 @@
+        .kill_on_drop(true)
+        .spawn()
+        .expect(&format!("{} failed to spawn", command_name))
+        .await?;
+
+    if !status.success() {
+        failure::bail!(
+            "tried running command:\n{}\nexited with {}",
+            command_name.replace("\"", ""),
+            status
+        )
+    }
+    Ok(())
+}
+
+async fn enable_tailing_start_heartbeat(
+    target: &Target,
+    user: &GlobalUser,
+) -> Result<(), failure::Error> {
+    let client = http::cf_v4_api_client_async(user, HttpApiClientConfig::default())?;
+
+    let url = get_tunnel_url().await?;
+
+    let response = client
+        .request(&CreateTail {
+            account_identifier: &target.account_id,
+            script_name: &target.name,
+            params: CreateTailParams { url },
+        })
+        .await;
+
+    match response {
+        Ok(success) => {
+            eprintln!("Now prepared to stream logs.");
+
+            let tail_id = success.result.id;
+
+            // Loop indefinitely to send "heartbeat" to API and keep log streaming alive.
+            loop {
+                thread::sleep(Duration::from_secs(60));
+                let heartbeat_result = send_heartbeat(target, &client, &tail_id).await;
+                if heartbeat_result.is_err() {
+                    return heartbeat_result;
+                }
+                // This should loop forever until SIGINT is issued or Wrangler process is killed
+                // through other means.
+            }
+        }
+        Err(e) => failure::bail!(http::format_error(e, None)),
+    }
+}
+
+async fn print_logs(req: Request<Body>) -> Result<Response<Body>, hyper::Error> {
+    match (req.method(), req.uri().path()) {
+        (&Method::POST, "/") => {
+            let whole_body = hyper::body::to_bytes(req.into_body()).await?;
+            println!("{}", str::from_utf8(&whole_body).unwrap());
+
+            Ok(Response::new(Body::from("Success")))
+        }
+        _ => {
+            let mut not_found = Response::default();
+            *not_found.status_mut() = StatusCode::NOT_FOUND;
+            Ok(not_found)
+        }
+    }
+}
+
+async fn get_tunnel_url() -> Result<String, failure::Error> {
+    // regex for extracting url from cloudflared metrics port.
+    let url_regex = Regex::new("userHostname=\"(https://[a-z.-]+)\"").unwrap();
+
+    let mut attempt = 0;
+
+    // This exponential backoff retry loop retries retrieving the cloudflared endpoint url
+    // from the cloudflared /metrics endpoint until it gets the URL or has tried retrieving the URL
+    // over 5 times.
+    while attempt < 5 {
+        if let Ok(resp) = reqwest::get("http://localhost:8081/metrics").await {
+            let body = resp.text().await?;
+
+            for cap in url_regex.captures_iter(&body) {
+                return Ok(cap[1].to_string());
+            }
+        }
+
+        attempt = attempt + 1;
+        thread::sleep(Duration::from_millis(attempt * attempt * 100));
```

Comment:
possibly worth adding a comment about why this sleep is necessary and how the value is constructed. seems like it gets exponentially longer as attempts increase? possible performance bottleneck

Reply:
BLAST FROM THE PAST ;)! This is a common technique called exponential backoff. It is standard for API calls (whenever a resource request fails and can be re-requested over the network, the time to re-request should exponentially increase to avoid creating packet receive queues which can cause network congestion/bottlenecks: https://en.wikipedia.org/wiki/Exponential_backoff)

Reply:
i believe it is covered in a comment earlier, it just requires you already know what exponential backoff is to tie the comment to the sleep call.

Reply:
TIL! Never knew about exponential backoff :) (hi gabbi)